src/freertos_drivers/common/SPIFlash.hxx

```diff
@@ -0,0 +1,168 @@
+/** \copyright
+ * Copyright (c) 2021, Balazs Racz
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are  permitted provided that the following conditions are met:
+ *
+ *  - Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  - Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file SPIFlash.hxx
+ *
+ * Shared implementation for operating spiflash devices. This class is intended
+ * to be used by other device drivers.
+ *
+ * @author Balazs Racz
+ * @date 4 Dec 2021
+ */
+
+#ifndef _FREERTOS_DRIVERS_COMMON_SPIFLASH_HXX_
+#define _FREERTOS_DRIVERS_COMMON_SPIFLASH_HXX_
+
+#include <inttypes.h>
+#include <spi/spidev.h>
+#include <sys/types.h>
+
+class OSMutex;
+class SPI;
+
+/// Create a const structure like this to tell the spiflash driver how to talk
+/// to your spiflash device.
+///
+/// Use it like this:
+///
+/// static const SPIFlashConfig cfg = {
+///    .speedHz_ = 2000000,
+///    .spiMode_ = 3,
+///    .writeCommand_ = 0xff,
+///};
+struct SPIFlashConfig
+{
+    /// Use this frequency to talk to SPI.
+    uint32_t speedHz_ {1000000};
+
+    /// How many bytes is an erase sector.
+    uint32_t sectorSize_ {4 * 1024};
```

Comment:
This value should come from a table based on reported ID. A todo and an issue filed for enhancement is good enough for now. Of course, we may need to get this value before instantiating the SPIFFS so that it gets the correct sector size too.

We may find ourselves down the road needing to have a second source for the SPI flash that uses different sector and page sizes.

Reply:
These are just the default initialization values. The expected usage is that there is a struct in HwInit which can be filled up for the given hardware. We could possibly put this into RAM and change the values, or select from more than one such structure that are in flash.

Reply:
Ultimately, this is discoverable at runtime by evaluating the SPI FLASH ID itself. That is how we should get the value.

Reply:
filed #586